docs/arch-split.md

```diff
@@ -512,12 +573,82 @@ global_interpretation Retype_AI?: Retype_AI
 end
 ```
 
-Note that the custom command `extend_locale` allows us to pretend that
-locales can be extended incrementally. This allows us to convert lemmas
-to locale assumptions in-place, without having to move locale
-assumptions to the point where the locale is initially declared.
+Note: `global_interpretation Retype_AI?` exports all names into the global
+context, but they are also accessible with a `Retype_AI.` prefix, which is
+useful to know in case of name collision.
+
+
+### Need for multiple theory-specific architecture-generic locales
 
-## Qualify
+In the previous section, we mentioned "locales" (plural), with the ideal case
+being a single locale. The reason we need this is due to chains of lemmas that
+look like this:
+* using AA (arch-specific) prove GB (generic)
+* using GB (generic) prove AC (arch-specific)
+* using AC (arch-specific) prove GD (generic)
+
+If we wanted to split this theory as in the single-locale example, we would end
+up with multiple theories, divided only by the arbitrariness of the dependency
+graph. Instead, we can employ multiple locales as follows.
+
+In the generic theory (e.g. `Foo_AI`) we do:
+
+```isabelle
+locale Foo_AI_1 =
+  assumes AA: (* ... *) "XXX"
+
+lemma (in Foo_AI_1) GB: "XXX"
+  by (* using AA *)
+
+locale Foo_AI2 = Foo_AI_1 +
+  assumes AC: (* ... *) "XXX"
+
+lemma (in Foo_AI_2) GD: "XXX"
+  by (* using AC *)
+```
+
+This allows instantiation of the locales in stages in the arch-specific theory
+(e.g. `ArchFoo_AI`):
+
+```isabelle
+
+context Arch begin global_naming ARM
+
+lemma AA[Foo_AI_asms]: "XXX"
+  by (* ... *)
+
+end
+
+interpretation Foo_AI_1?: Foo_AI_1
+  proof goal_cases
+  interpret Arch .
+  case 1 show ?case by (intro_locales; (unfold_locales; fact Foo_AI_asms)?)
+  qed
+
+(* now we get access to GB in the global context *)
+
+context Arch begin global_naming ARM
+
+lemma AC[Foo_AI_asms]: "XXX"
+  by (* using GB *)
+
+end
+
+interpretation Foo_AI_2?: Foo_AI_2
+  proof goal_cases
+  interpret Arch .
+  case 1 show ?case by (intro_locales; (unfold_locales; fact Foo_AI_asms)?)
+  qed
+
+(* now we get access to GD in the global context *)
+```
+
+FIXME: strictly speaking, there is no requirement that `Foo_AI_2` includes
+`Foo_AI_1`, but in practice we always do so (I guess because we want everything
+that got proved to be instantly available in case we need it in the proof
+chain). This would mean that the same fact gets exported multiple times?
```

Comment:
I'm not entirely sure, but I think Isabelle is actually smart about that. We could double check, but the number of lemmas that would be instantiated twice is fairly small, so I would probably stick with a chain like this even if there was duplication.

Reply:
Yeah, I guess I should check.

- [ ] investigate how many lemmas of same name but different prefix this chain generates

Reply:
Investigation results: at most one duplicate is generated, so can confirm that Isabelle is smart about it as you said. My initial thought that maybe we can make all the locales export with the same prefix has been demonstrated to be impossible as well, so we can leave the status quo as is, and I should remove this FIXME. Maybe I should resolve it into a note in the doc? Would that be of use to anyone?

Reply:
If it can be expressed easily, it would be of use in avoiding us having the same thought and doing the same investigation some time in the future.

Reply:
Agreed, adding a note would be nice, I'm sure we'll have the same question again in a few years otherwise.